### PR TITLE
Add a c-api interface to initialize the thread environment of Paddle

### LINKED ISSUE #5669: C-API库进行gpu多线程inference时，出现cublas status: not initialized错误
从网站上下载的cuda 8.0的libpaddle_capi_shared.so, 使用一个全连接的网络，在进行forward时，出现如下错误：
```text
F1115 15:36:56.346094 17370 hl_cuda_cublas.cc:307] Check failed: stat == CUBLAS_STATUS_SUCCESS (1 vs. 0) [cublas status]: not initialized
*** Check failure stack trace: ***
    @     0x7f8ddce75bcd  google::LogMessage::Fail()
    @     0x7f8ddce7967c  google::LogMessage::SendToLog()
    @     0x7f8ddce756f3  google::LogMessage::Flush()
    @     0x7f8ddce7ab8e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f8ddd265147  hl_matrix_mul()
    @     0x7f8ddce7ab8e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f8ddd265147  hl_matrix_mul()
    @     0x7f8ddd0c3987  paddle::GpuMatrix::mul()
    @     0x7f8ddd0c3987  paddle::GpuMatrix::mul()
    @     0x7f8ddd0c3fe1  paddle::GpuMatrix::mul()
    @     0x7f8ddcfc5ffb  paddle::FullyConnectedLayer::forward()
    @     0x7f8ddd0c3fe1  paddle::GpuMatrix::mul()
    @     0x7f8ddcfc5ffb  paddle::FullyConnectedLayer::forward()
    @     0x7f8ddcedf95d  paddle::NeuralNetwork::forward()
    @     0x7f8ddce6f6b6  paddle_gradient_machine_forward
    @     0x7f8ddcedf95d  paddle::NeuralNetwork::forward()
    @           0xc52da7  recarch::paddle::PaddlePredictor::predict()
    @     0x7f8ddce6f6b6  paddle_gradient_machine_forward
    @           0x7e20f1  rec::predictor::CtrDnnEngineV4::calc_quality_callback()
    @           0xc52da7  recarch::paddle::PaddlePredictor::predict()
    @           0x7998a1  rec::predictor::BaseEngine::handle()
    @           0x79e646  rec::predictor::PredictTask::run()
    @           0x7e20f1  rec::predictor::CtrDnnEngineV4::calc_quality_callback()
    @           0x7a63cd  rec::predictor::WorkerThread::run()
    @           0x7998a1  rec::predictor::BaseEngine::handle()
    @           0x79e646  rec::predictor::PredictTask::run()
    @           0xf4c6aa  thread_proxy
    @           0x7a63cd  rec::predictor::WorkerThread::run()
    @           0xf4c6aa  thread_proxy
    @     0x7f8de08021c3  start_thread
    @     0x7f8de08021c3  start_thread
    @     0x7f8ddc0da12d  __clone
    @     0x7f8ddc0da12d  __clone
    @              (nil)  (unknown)
```
请问这个是什么原因？怎么解决？

hl_matrix_mul函数报错代码在：
    stat = CUBLAS_GEMM(t_resource.handle,
                       CUBLAS_OP_N,
                       CUBLAS_OP_N,
                       dimN,
                       dimM,
                       dimK,
                       &alpha,
                       B_d,
                       ldb,
                       A_d,
                       lda,
                       &beta,
                       C_d,
                       ldc);
其中t_resource.handle是一个空指针，原因是t_resource是一个thread_local的变量，没有初始化。

### PR DESCRIPTION
**Pull request by Xreki**
- Fix #5669 
  - `paddle_init_thread()` need to be called in a newly launched thread to initialize the thread environment of Paddle
- Add a GPU c-api example

### CODE REVIEW

**Comment by QingshuChen** [comment]
hl_init() will set the t_resource.device to -1, is it a bug?

File: paddle/capi/Main.cpp [hunk 1]
```diff
@@ -43,4 +43,16 @@ paddle_error paddle_init(int argc, char** argv) {
   isInit = true;
   return kPD_NO_ERROR;
 }
+
+paddle_error paddle_init_thread() {
+  static __thread bool isInit = false;
+  if (isInit) return kPD_NO_ERROR;
+
+  if (FLAGS_use_gpu) {
+    hl_init(FLAGS_gpu_id);
```
Comment (QingshuChen) [inline paddle/capi/Main.cpp]: hl_init will set the t_resource.device to -1, is it a bug?

File: paddle/capi/error.h [hunk 2]
```diff
@@ -27,4 +27,21 @@ typedef enum {
   kPD_UNDEFINED_ERROR = -1,
 } paddle_error;
 
+static const char* paddle_error_string(paddle_error err) {
```
Comment (hedaoyuan) [inline paddle/capi/error.h]: Move the function implementation into a .cc file and avoid generating duplicate copies of code in the inference library.

File: paddle/capi/main.h [hunk 3]
```diff
@@ -26,6 +26,11 @@ extern "C" {
  */
 PD_API paddle_error paddle_init(int argc, char** argv);
 
+/**
+ * Initialize the thread environment of Paddle.
+ */
+PD_API paddle_error paddle_init_thread();
```
Comment (hedaoyuan) [inline paddle/capi/main.h]: Since some previous CPU users did not use this interface. Here need to indicate only GPU need this interface.

**Review by hedaoyuan — CHANGES_REQUESTED** [review]

File: paddle/capi/Main.cpp [hunk 4]
```diff
@@ -43,4 +43,16 @@ paddle_error paddle_init(int argc, char** argv) {
   isInit = true;
   return kPD_NO_ERROR;
 }
+
+paddle_error paddle_init_thread() {
```
[the remark this answers is not in the document]
Comment (QingshuChen) [reply paddle/capi/Main.cpp]: If adding a device_id, another question is when user invokes the paddle_matrix_create API, which gpu is the matrix on? There is no device id the the paddle_matrix_create API.

[the remark this answers is not in the document]
Comment (hedaoyuan) [reply paddle/capi/Main.cpp]: Maybe need to wrap the `hl_get_device` interface.

[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/capi/Main.cpp]: I think @QingshuChen is right. How to change the `device_id` of `Matrix`?

[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/capi/Main.cpp]: I tried to add `device_id` and set it to different values. The binary failed:
```
F1120 08:05:09.430253  1282 hl_cuda_device.cc:565] Check failed: cudaSuccess == cudaStat (0 vs. 77) Cuda Error: an illegal memory access was encountered
```

File: paddle/capi/Main.cpp [hunk 5]
```diff
@@ -43,4 +43,16 @@ paddle_error paddle_init(int argc, char** argv) {
...
+  static __thread bool isInit = false;
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/capi/Main.cpp]: Removed.

File: paddle/capi/error.h (hunk 2 above)
Comment (Xreki) [reply paddle/capi/error.h]: Done.

File: paddle/capi/main.h (hunk 3 above)
Comment (Xreki) [reply paddle/capi/main.h]: Done.

File: paddle/capi/Main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (hedaoyuan) [reply paddle/capi/Main.cpp]: This problem is due to the thread share the same parameter model with the main thread. In this case, you need to pass in the same device id as the main thread.

[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/capi/Main.cpp]: I think I know what your means. In the current mode, the `Parameters` are shared among multi-threads, but each thread has its own `network`. If each thread wants to run on different GPU, it should have its own `network` + `Parameters`.  Then `Parameters` cannot be shared anymore, and we should not use the interface `paddle_gradient_machine_create_shared_param` but directly using `paddle_gradient_machine_create_for_inference` and `paddle_gradient_machine_load_parameter_from_disk` instead.

[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/capi/Main.cpp]: - 我认为，这个多线程的例子，是希望使用`paddle_gradient_machine_create_shared_param`共享参数，不能支持跑在不同的GPU上。
- 如果希望多个线程跑在不同的GPU上，则应该使用`trainer_count>1`或者，在每个线程里面单独地`paddle_gradient_machine_create_for_inference`以及`paddle_gradient_machine_load_parameter_from_disk`。

File: paddle/capi/examples/model_inference/multi_thread/main_gpu.c [hunk 6]
```diff
@@ -0,0 +1,106 @@
+#include <paddle/capi.h>
+#include <pthread.h>
+#include <time.h>
+#include "../common/common.h"
+
```
Comment (hedaoyuan) [inline paddle/capi/examples/model_inference/multi_thread/main_gpu.c]: Add some comments about this example. For example, this is an inference implementation where multiple threads share a GPU.

**Review by hedaoyuan — DISMISSED** [review]

**Review by hedaoyuan — DISMISSED** [review]

File: paddle/capi/examples/model_inference/multi_thread/main_gpu.c (hunk 6 above)
Comment (Xreki) [reply paddle/capi/examples/model_inference/multi_thread/main_gpu.c]: Done.

**Review by hedaoyuan — APPROVED** [review]

**Pull request merged**